src/estimation/src/AttitudeQuaternionEKF.cpp

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright (C) 2019 Fondazione Istituto Italiano di Tecnologia
+ *
+ * Licensed under either the GNU Lesser General Public License v3.0 :
+ * https://www.gnu.org/licenses/lgpl-3.0.html
+ * or the GNU Lesser General Public License v2.1 :
+ * https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * at your option.
+ */
+
+#include <iDynTree/Estimation/AttitudeQuaternionEKF.h>
+
+void iDynTree::AttitudeQuaternionEKF::serializeStateVector()
+{
+    if (m_x.size() != m_state_size)
+    {
+        m_x.resize(m_state_size);
+    }
+
+    iDynTree::toEigen(m_x).block<4, 1>(0, 0) = iDynTree::toEigen(m_state.m_orientation);
+    iDynTree::toEigen(m_x).block<3, 1>(4, 0) = iDynTree::toEigen(m_state.m_angular_velocity);
+    iDynTree::toEigen(m_x).block<3, 1>(7, 0) = iDynTree::toEigen(m_state.m_gyroscope_bias);
+}
+
+void iDynTree::AttitudeQuaternionEKF::deserializeStateVector()
+{
+    iDynTree::toEigen(m_state.m_orientation) = iDynTree::toEigen(m_x).block<4, 1>(0, 0);
+    iDynTree::toEigen(m_state.m_angular_velocity) = iDynTree::toEigen(m_x).block<3, 1>(4, 0);
+    iDynTree::toEigen(m_state.m_gyroscope_bias) = iDynTree::toEigen(m_x).block<3, 1>(4, 0);
+}
+
+void iDynTree::AttitudeQuaternionEKF::prepareSystemNoiseCovarianceMatrix(iDynTree::MatrixDynSize &Q)
+{
+    if (Q.rows() != m_state_size && Q.cols() != m_state_size)
+    {
+        Q.resize(m_state_size, m_state_size);
+    }
+
+    iDynTree::MatrixDynSize Fu(m_state_size, m_input_size + m_state.m_gyroscope_bias.size());
+    Fu.zero();
+    iDynTree::toEigen(Fu).block<3, 3>(4, 0) = iDynTree::toEigen(m_Id3);
```

Comment:
I see a lot of `iDynTree::toEigen(m_Id3)` around in the code. Perhaps `Matrix<double, 3, 3>::Identity()` is a bit more idiomatic and more readable? But it is your call.

Reply:
I'd prefer m_Id3 and documenting it properly. Since the alternative will make the statements longer and less readable, imo (in the sense, it will be distracting).